New ratelimiting implementation

davfsa opened the pull request:
Hey!

This pull request adds support for nirn to utilize a sliding window technique with regards to ratelimits, which allows for a higher throughput and (optionally, which can be enabled through `ALLOW_CONCURRENT_REQUESTS=true`) the ability to perform concurrent requests to the same bucket!

I also took the liberty to update all the dependencies to their latest version, as of writing, and to update the usage of some deprecated apis.

davfsa committed:
Implement sliding window ratelimiting

davfsa committed:
Raise ctx error when acquiring

davfsa committed:
Cleanup logs

davfsa committed:
Add config option to disable concurrent requests

davfsa committed:
Improve bucket logging

davfsa committed:
Improve timings

davfsa committed:
Fix handling of global ratelimits

davfsa committed:
Finish off

davfsa committed:
Improve err log

davfsa committed:
Update documentation on new option

davfsa committed:
Use mutex instead of RWMutex for queue channels

davfsa committed:
Adjust lock when doing a bucket update

davfsa committed:
Revert incorrect log change

davfsa committed:
Remove healing log

davfsa commented:
Some examples:

https://github.com/user-attachments/assets/cf63dd73-ea76-4f32-8c7c-df24242fef2e

<details>
<summary>Script</summary>

```py
from __future__ import annotations

import asyncio
import os

import dotenv

import hikari

dotenv.load_dotenv()

CHANNEL_ID = 

bot = hikari.GatewayBot(
    logs={
        "version": 1,
        "incremental": True,
        "loggers": {"": {"level": "INFO"}, "hikari.ratelimits": {"level": "DEBUG"}},
    },
    rest_url="http://localhost:8080/api/v10",
    max_retries=0,
    cache_settings=hikari.impl.CacheSettings(
        components=hikari.impl.CacheComponents.NONE
    ),
    token=os.environ["TOKEN"],
)


async def spam() -> None:
    while True:
        await bot.rest.create_message(CHANNEL_ID, "Hello!")


@bot.listen()
async def started(_: hikari.StartedEvent) -> None:
    await asyncio.gather(*(spam() for _ in range(5)))


bot.run()
```

</details>

---

https://github.com/user-attachments/assets/82b827a5-33bf-4c65-855e-77b312596045

<details>
<summary>Script</summary>

```py
from __future__ import annotations

import asyncio
import os

import dotenv

import hikari

dotenv.load_dotenv()

bot = hikari.GatewayBot(
    logs={
        "version": 1,
        "incremental": True,
        "loggers": {"": {"level": "INFO"}, "hikari.ratelimits": {"level": "DEBUG"}},
    },
    rest_url="http://localhost:8080/api/v10",
    max_retries=0,
    cache_settings=hikari.impl.CacheSettings(
        components=hikari.impl.CacheComponents.NONE
    ),
    token=os.environ["HIKARI_TOKEN"],
)

CHANNEL_ID =
BOT_ID =
message_id = 0


async def spam() -> None:
    while True:
        await bot.rest.add_reaction(CHANNEL_ID, message_id, "\u2764\ufe0f")
        await bot.rest.delete_my_reaction(CHANNEL_ID, message_id, "\u2764\ufe0f")


@bot.listen()
async def test(event: hikari.ReactionAddEvent) -> None:
    if event.message_id != message_id or event.user_id == BOT_ID:
        return

    if event.emoji_id:
        await bot.rest.delete_reaction(
            event.channel_id,
            event.message_id,
            event.user_id,
            emoji=event.emoji_name,
            emoji_id=event.emoji_id,
        )
    else:
        await bot.rest.delete_reaction(
            event.channel_id, event.message_id, event.user_id, emoji=event.emoji_name
        )


@bot.listen()
async def started(_: hikari.StartedEvent) -> None:
    message = await bot.rest.create_message(
        CHANNEL_ID, "Spam reactions on me, I will delete them :D"
    )
    await message.add_reaction("\u2764")

    global message_id
    message_id = message.id

    asyncio.create_task(spam())


bot.run()
```

</details>

---

*Note:, hikari was patched to comment out these two lines https://github.com/hikari-py/hikari/blob/f07f4187b9e6d284e334cb913625b2988217a13f/hikari/impl/rest.py#L802-L803 to allow nirn to fully take over handling ratelimiting*

davfsa committed:
Improve parsing of headers

davfsa committed:
Add some logging information

davfsa committed:
Fix incorrect negation

davfsa committed:
Use identifier instead of user ID

davfsa committed:
Remove extra logging

davfsa committed:
Get scope from correct header

davfsa committed:
Lower remaining too if limit is lowered

davfsa committed:
Remove unused argument

davfsa commented on lib/queue.go:
```diff
@@ -254,10 +269,9 @@ func return404webhook(item *QueueItem) {
 	if err != nil {
 		item.errChan <- err
-		return
+	} else {
+		item.doneChan <- nil
 	}
```
```suggestion
		return
	}
	item.doneChan <- nil
```

davfsa committed:
Improve ratelimit handling when starting already hitting ratelimits or starting in the middle of a window

This can help in cases where NIRN is restarted and it loses all ratelimiting information

davfsa committed:
Re-add case for slide period differing too much with adjusted limits

davfsa committed:
Update logging wording

davfsa committed:
fix: add support for routes that have a fixed window

Signed-off-by: davfsa <davfsa@gmail.com>

davfsa committed:
feat: improve detection of fixed window ratelimits

Signed-off-by: davfsa <davfsa@gmail.com>

davfsa committed:
feat: improve detection of fixed buckets

davfsa committed:
fix: typo

davfsa committed:
feat: greatly improve ratelimit avoidance by tracking in transit requests

davfsa committed:
chore: drop logging level

davfsa committed:
chore: readd self-healing logic

davfsa committed:
chore: bump go version

davfsa committed:
chore: update bucket name when updating ratelimiter

davfsa committed:
chore: improve struct alignment

germanoeich commented:
@codex deep review

davfsa commented on lib/ratelimits.go:
```diff
@@ -0,0 +1,262 @@
+	return b.remaining <= 0
+}
+
+// Acquire will request a slot from the ratelimit and sleep until there is one available
+func (b *BucketRateLimit) Acquire(ctx context.Context) error {
+	b.inTransitLock.Lock()
+	if b.inTransit >= b.limit {
+		b.transitWaitChan = make(chan interface{})
+		b.inTransitLock.Unlock()
+		select {
+		case <-ctx.Done():
```
Damn, nice catch

davfsa committed:
fix: fix issues from review

davfsa commented:
@germanoeich All fixed :)

davfsa commented on lib/bucket.go:
```diff
@@ -0,0 +1,270 @@
+			gain := int64(math.Floor((now.Sub(b.increaseAt).Seconds())/b.period.Seconds())) + 1
+			nowRemaining := b.remaining + gain
+
+			b.remaining = min(nowRemaining, b.limit)
+
```
min and max are builtins in go as of go 1.21

germanoeich replied:
`You're absolutely right!` god fucking damn I can smell claude from a mile away.

davfsa commented on lib/ratelimits.go:
```diff
@@ -0,0 +1,270 @@
+		b.inTransit++
+		b.inTransitLock.Unlock()
+	}
+
```
This is intended to support only a single concurrent request by design

germanoeich commented:
@davfsa nice, don't pay much attention to the code rabbit comments btw, not a fan of their reviews and will likely remove it later. Anyway thanks for the fixes, I'll take a final look over this soon

davfsa commented on lib/ratelimits.go:
```diff
@@ -0,0 +1,270 @@
+		b.limit = limit
+		b.remaining = min(b.remaining, b.limit)
+	}
+
```
This would cause too much hassle and state to keep in sync. The code is meant to handle an upwards and downwards update of limit correctly, as later on Acquire will still check it properly.

davfsa committed:
feat: add anti-deadlock measures

davfsa committed:
fix: add missing return

davfsa committed:
chore: do not shadow err variable

davfsa committed:
feat: improve global locked until sleep

davfsa committed:
chore: cleanup locks and add some documentation

germanoeich commented:
@codex deep review

germanoeich commented on lib/queue.go:
```diff
@@ -282,106 +297,150 @@ func isInteraction(url string) bool {
-					"until":      time.Now().Add(resetAfter),
-					"resetAfter": resetAfter,
-				}).Warn("Global reached, locking")
-			}
+		ch.Lock()
+		if ch.lockerFun != nil {
+			ch.lockerFun(item)
+			continue
```
@davfsa can you take a look at this one? After this I'll be merging :pray:

davfsa replied:
Thank you, nice catch!

Will modify

davfsa replied:
@germanoeich Updated!

davfsa committed:
fix: add missing unlock to prevent a deadlock

davfsa commented:
@codex deep review

Riksou commented:
@codex deep review

davfsa commented on lib/ratelimits.go:
```diff
@@ -0,0 +1,274 @@
+	return b.remaining <= 0
+}
+
+// Acquire will request a slot from the ratelimit and sleep until there is one available
+func (b *BucketRateLimit) Acquire(ctx context.Context) error {
+	b.inTransitLock.Lock()
+	if b.inTransit >= b.limit {
+		// Buffer of 1 here to prevent deadlocks in a worst case scenario
+		b.transitWaitChan = make(chan interface{}, 1)
+		b.inTransitLock.Unlock()
+		select {
+		case <-ctx.Done():
+			b.inTransitLock.Lock()
+			if b.transitWaitChan != nil {
+				b.transitWaitChan = nil
+			} else {
+				// Return the slot that was given to us
+				b.inTransit--
+			}
+			b.inTransitLock.Unlock()
+			return ctx.Err()
+		case <-b.transitWaitChan:
+		}
```
This is expected, as multiple concurrent calls to Acquire are not supported.

Support can be made for it by using a proper semaphore, but for now I have decided the complexity is not worth it

davfsa committed:
fix: release bucket before erroring when context is done

davfsa committed:
fix: add more precise lock section

davfsa committed:
chore: make ratelimiter never nil

davfsa committed:
chore: microoptimization

davfsa commented on lib/ratelimits.go:
```diff
@@ -0,0 +1,292 @@
+func (b *BucketRateLimit) isRatelimited(now time.Time) bool {
+	if b.unknown {
+		// Don't do any waiting logic as we don't have any information on the bucket,
+		// just do the request immediately
```
```suggestion
		// Don't do any waiting logic as we don't have any information on the bucket,
		// just do the request immediately. The first successful request will return the bucket
		// data
```

davfsa committed:
Apply suggestion from @davfsa

davfsa committed:
fix: improve logic to detect fixed buckets

davfsa committed:
fix: decrease sliding precision

davfsa committed:
fix: proper order of logic

davfsa committed:
feat: improve detection of fixed buckets

Assume they will never change from fixed to sliding.

This can be massively cleaned up if Discord just exposed the information

Signed-off-by: davfsa <davfsa@gmail.com>

davfsa committed:
Revert "feat: improve detection of fixed buckets"

This reverts commit 649bfb0b0745083f92a15ae5ff5073dd95b773bf.

davfsa committed:
chore: treat bucket as fixed by default

Should help in cases where the bucket is hit in the middle of a slide, like when nirn is restarted in the middle of it's runtime with pending requests

davfsa committed:
chore: only update bucket type if not first ratelimit information

davfsa commented:
Going to freeze this PR in this state.

We have deployed it at Invite Tracker and it seems to work splendidly for all routes and allows us a higher throughput thanks to parallel requests!

If anybody deploys this at a large scale or finds issues with it, please reach out to me on Discord :)

davfsa committed:
feat: improve 429 handling

davfsa committed:
chore: drop logging level

davfsa committed:
chore: readd math.Ceil call

davfsa committed:
fix: only re-initialize bucket if limit changed

davfsa committed:
chore: remove redundant parenthesis

davfsa committed:
chore: cleanup logging

davfsa committed:
chore: re-add self healing logic

davfsa committed:
fix: make bucket following less strict for insanely fast connections

This can happen due to all Discord responses having a slight margin of error, which makes it likely for ratelimits to be hit if you are *too* quick

davfsa committed:
chore: remove random spaces

davfsa committed:
chore: remove incorrect optimistic bucket path

davfsa committed:
chore: remove sliding prevention if out of sync

davfsa committed:
chore: re-add after or equal check

davfsa committed:
chore: speedup arm docker builds

Signed-off-by: davfsa <davfsa@gmail.com>

davfsa committed:
chore: also build docker images on pushes to branches

Signed-off-by: davfsa <davfsa@gmail.com>

davfsa committed:
fix: go build instead of go install

Signed-off-by: davfsa <davfsa@gmail.com>

davfsa committed:
chore: re-disable ci in testing branches

Signed-off-by: davfsa <davfsa@gmail.com>

davfsa committed:
fix: properly calculate message delete bucket path

davfsa committed:
feat: add more logging to buckets

davfsa committed:
fix: properly handle route scoped ratelimit hits

davfsa committed:
fix: drop logger back to debug

davfsa committed:
feat: multi-bucket handling (#1)

davfsa committed:
test: allow sliding window even if out of sync

davfsa committed:
fix: use proper parameter for buckets sweep

davfsa committed:
fix: remove test code

davfsa committed:
chore: account for network delay

davfsa committed:
chore: cleanup

davfsa committed:
chore: remove unused field

davfsa committed:
fix: remove incorrect assumption of major channels bucket

davfsa committed:
feat: improve logging

davfsa committed:
feat: enable concurrent requests by default

davfsa committed:
feat: improve handling of buckets with top level resources and multiple sub-buckets

davfsa committed:
fix: properly group up CF per route ban

davfsa committed:
chore: group up all interaction callbacks into a single queue

davfsa committed:
feat: improve logging

davfsa committed:
fix: avoid spamming metrics with activity-instance IDs

davfsa committed:
fix: do not distribute interaction callbacks over cluster

davfsa committed:
fix: do not merge callback endpoints

It causes issues that are too rooted into how endpoints are handled, so just split them up for now

davfsa committed:
chore: bump deps

davfsa committed:
fix: detection of fixed buckets

davfsa committed:
chore: small optimization

davfsa committed:
chore: revert some old behavior and prevent period increase

davfsa committed:
chore: prevent resetAt and increaseAt from sliding backwards

davfsa committed:
fix: prevent path traversal at proxy level

davfsa committed:
chore: use 422 instead of 400 for path traversal

davfsa committed:
fix: increase inactive bucket sweep time + properly calculate bucket first valid headers check

davfsa committed:
fix: do not drop old ratelimit information

davfsa committed:
chore: cleanup bucket last updated at

davfsa committed:
feat: add nanosecond precision to log messages

davfsa committed:
fix: avoid immediately collecting newly created buckets

davfsa committed:
fix: reduce number of 429s

davfsa committed:
fix: revert incorrect change

Signed-off-by: davfsa <davfsa@gmail.com>
